core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java

```diff
@@ -58,40 +60,83 @@
   static final String RENAME_OPERATION_DESCRIPTION = "Renamed as a part of a join";
 
   private final JoinerConfig conf;
-  private Schema outputSchema;
-  private Map<String, List<String>> perStageJoinKeys;
-  private Table<String, String, String> perStageSelectedFields;
-  private JoinConfig joinConfig;
-  private Map<String, Schema> keySchemas = new HashMap<>();
 
   public Joiner(JoinerConfig conf) {
     this.conf = conf;
   }
 
+  @Nullable
   @Override
-  public void configurePipeline(MultiInputPipelineConfigurer pipelineConfigurer) {
-    MultiInputStageConfigurer stageConfigurer = pipelineConfigurer.getMultiInputStageConfigurer();
-    Map<String, Schema> inputSchemas = stageConfigurer.getInputSchemas();
-    FailureCollector collector = pipelineConfigurer.getMultiInputStageConfigurer().getFailureCollector();
-    init(inputSchemas, collector);
-    collector.getOrThrowException();
-    if (!conf.inputSchemasAvailable(inputSchemas) && !conf.containsMacro(JoinerConfig.OUTPUT_SCHEMA) &&
+  public JoinDefinition define(AutoJoinerContext context) {
+    FailureCollector collector = context.getFailureCollector();
+
+    boolean hasUnknownInputSchema = context.getInputStages().values().stream().anyMatch(Objects::isNull);
+    if (hasUnknownInputSchema && !conf.containsMacro(JoinerConfig.OUTPUT_SCHEMA) &&
       conf.getOutputSchema(collector) == null) {
       // If input schemas are unknown, an output schema must be provided.
       collector.addFailure("Output schema must be specified", null).withConfigProperty(JoinerConfig.OUTPUT_SCHEMA);
     }
 
-    Schema outputSchema = getOutputSchema(inputSchemas, collector);
-    if (outputSchema != null) {
-      // Set output schema if it's not a macro.
-      stageConfigurer.setOutputSchema(outputSchema);
+    if (conf.requiredPropertiesContainMacros()) {
+      return null;
+    }
+
+    Set<String> requiredStages = conf.getRequiredInputs();
+    Set<String> broadcastStages = conf.getBroadcastInputs();
+    List<JoinStage> inputs = new ArrayList<>(context.getInputStages().size());
+    boolean useOutputSchema = false;
+    for (JoinStage joinStage : context.getInputStages().values()) {
+      inputs.add(JoinStage.builder(joinStage)
+        .setRequired(requiredStages.contains(joinStage.getStageName()))
+        .setBroadcast(broadcastStages.contains(joinStage.getStageName()))
+        .build());
+      useOutputSchema = useOutputSchema || joinStage.getSchema() == null;
+    }
+
+    try {
+      JoinDefinition.Builder joinBuilder = JoinDefinition.builder()
+        .select(conf.getSelectedFields(collector))
+        .from(inputs)
+        .on(JoinCondition.onKeys()
+              .setKeys(conf.getJoinKeys(collector))
+              .setNullSafe(conf.isNullSafe())
+              .build());
+      if (useOutputSchema) {
+        joinBuilder.setOutputSchema(conf.getOutputSchema(collector));
+      } else {
+        joinBuilder.setOutputSchemaName("join.output");
```

Comment:
Just curious - is the schema name useful for the output schema? I believe the builder will use its own schema name if this is not provided, and because different plugins seem to give different schema name, and the name basically is not used anywhere. Is it important and worth to pay attention when developing a plugin in the future?

Reply:
it's important with certain types of schemas that have recursive structures, because the schema name is used instead of the object schema to prevent infinite recursion. In most situations it doesn't matter too much, setting this here since this is the schema name used before.